### ISSUE TITLE
work with jsdom in node

### ISSUE THREAD

**Issue by taoqf** [dojo/dom#40]
i wish i could use it in node as well

**Comment by mwistrand** [comment]
@taoqf, are you suggesting that `dojo/dom` should work with [jsdom](https://github.com/tmpvar/jsdom), or that `dojo/dom` should have its own Node implementation of the DOM? If the former, I do not see how `dojo/dom` is not already compatible with jsdom.

**Comment by kitsonk** [comment]
@mwistrand when I read this, the only thing I am thinking we should do, is like we have done with `dojo/parser` (and we are going to do with `dojo/widgets`) is have the unit tests under node load JSDom, so the unit tests can be run "headless".

**Comment by kfranqueiro** [comment]
What would be the point of running unit tests against jsdom?  It does not represent an actual realistic browser environment.  (It's like the same age-old argument we've had against testing with PhantomJS.)

I mean, at the very least, I would not want that to lead to us "fixing" things that are only "broken" against a fake environment.

**Comment by dylans** [comment]
@kfranqueiro From my perspective, it is more important to make sure that the feature tests we write don't assume that we can predict every possible environment that is to come. There are certainly uses for Phantom2 and jsdom and similar libraries (server-side PDF generation, pre-rendering, etc.), and really there's little we're doing in this package that should not "just work" in both of those environments. I agree that if that's not the case, it's not worth investing our time currently to fix things against that environment.

A fairly big mistake we made with Dojo 1.0 (in my opinion) was assuming that we could predict the future and know all possible environments in which Dojo would run in. I don't want to repeat that mistake! :)

**Comment by vansimke** [comment]
Maybe I'm reading this wrong, but this appears to be a request for a change to Intern, not dojo-dom. As I understand it, the tests just run in whatever environment Intern gives it, whether that be node, a browser, or a CI provider. Doesn't that mean that the ability to provide a dom implementation for dojo-dom when running on node needs to come from it?

**Comment by kitsonk** [comment]
@vansimke it isn't changing Intern at all.

An example of how you can [load JSDom](https://github.com/dojo/widgets/blob/master/tests/support/loadJsdom.ts) [dynamically](https://github.com/dojo/widgets/blob/master/tests/unit/projector.ts#L1) when you want to run unit tests that require a DOM under node.

**Comment by vansimke** [comment]
Thanks @kitsonk, I'll try to apply this pattern and see how it goes.

**Comment by vansimke** [comment]
I have incorporated the JSDOM code that @kitsonk pointed me to and everything is working now. The only issue is, due to JSDOM's incomplete implementation of the standard, 45 of the 129 tests are failing. At this point, I can see two options:
1) check  this code in as-is and let the failing tests warn the user about where things won't work or,
2) update the code base to shim the parts of JSDOM that aren't working.

I would propose a hybrid of those two paths. For now, I would follow path (1) and just let developers know what works and what doesn't. I would then create a new issue to address the failing tests so that it can be prioritized along with the other issues.

Thoughts?

**Comment by kfranqueiro** [comment]
Given what you've found in terms of jsdom's incomplete implementation, I'd wonder what @dylans thinks given his earlier statement:

> I agree that if that's not the case, it's not worth investing our time currently to fix things against that environment.

Personally I'm -1 on checking in failing tests for not-really-a-real-DOM; I am also -1 on spending effort and adding code to "fix" it; given the number of packages we're working on and will need to maintain, I'd rather we pick our battles wisely.

**Comment by vansimke** [comment]
Well, to be fair, the tests already fail under intern:node, just much earlier since there isn't even a hint of a DOM-like structure. This code would just show where things stand when node is augmented by jsdom.

**Comment by kfranqueiro** [comment]
Er, yeah, I imagine they would fail under node right now.  I didn't really expect us to be running them directly in node to begin with.

**Comment by vansimke** [comment]
Exactly, by adding in the code that loads JSDOM in a node environment, then I think we would help others to know how well this module would work if they had a use case for it. I could imagine that it would in useful information, but I agree that getting it to work perfectly doesn't make sense. If anything, it might be good to use the failing tests as a source for creating issues for the JSDOM repo since they are looking for information like this.

**Comment by kitsonk** [comment]
My opinion is that the only thing more to expend effort on at the moment is to do a simple triage of what is missing from JSDom that is breaking the tests.  If it is something like a code path breaking because of a mis-identified feature, it maybe worth making changes in the library so that the code is more isomorphic and help harden any feature detection.  If it is several core fundamental features, then I suspect there is no value in getting `dojo/dom` running with JSDom.

We have `dojo/parser` and `dojo/widgets` that successfully run all their unit tests against JSDom without any special pathing.  I still argue that opens up opportunities for use of our libraries we might not envision.  In fact for `dojo/widgets` @pottedmeat and I have often discussed an approach to be able to generate static DOM server side for Dojo 2 applications that a) makes SEO possible with a Dojo 2 application and b) provide easy to develop progressive enhancement for Dojo 2 applications, improving user perceived load times and responsiveness.

**Comment by vansimke** [comment]
Okay, I'm not using the latest version of JSDOM since the code I stole from @kitsonk was for an older version. I'll move up to the latest version and get things working as best I can and then post a summary of the issues that I'm seeing.

**Comment by kitsonk** [comment]
I was using the older version because I was having some issues using more recent versions and it was working, so I pinned it there.

**Comment by vansimke** [comment]
Probably the same ones I was having. Still, it is a new major version, so I'd like to give JSDOM the best chance to implement features that I can.

**Comment by vansimke** [comment]
I've finished reviewing the current test suite when run on node with JSDOM and have identified the following issues:
* JSDOM has little to now support for CSS stylesheets so almost all of addCssRule's fail. This shouldn't be a show-stopper since any kind of isomorphic use of dojo-dom would not only need to ensure that the sheets get generated and added to the document, which could be proved in other ways
* dom works, except for the Node and Range objects aren't defined in this environment. Shims would have to be put in place to in make this work properly, specifically the has('dom-contextual-fragment') check in has.ts would have to be adjusted and Node.DOCUMENT_POSITION_CONTAINS would have to be provided for to make the contains() function work.
* form elements can't be accessed via the ['name'] syntax used in the tests in form.ts, these would have to be restructured to access the elements in a different way
* checkbox elements in JSDOM do not have a default value of "on", so some of the tests for form.select* would have to be reworked
* JSDOM does not appear to support the 'multiple' attribute of select tags properly
* all the tests work for delegate and schedule with no changes

In conclusion, I think that nothing here is insurmountable, except for the CSS items, which would probably make sense to ignore. What do you think @kitsonk? Should we go for it?

**Comment by kitsonk** [comment]
@vansimke yes, I would proceed on skipping those tests.  I think this is a lower priority though.  I had already set of milestone of June and I think we should put it on hold and revisit it in a while.
